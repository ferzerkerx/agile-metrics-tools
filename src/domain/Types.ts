import { MetricItem } from './metrics';

export interface TeamMetricsRequest {
  since?: Date;
  until?: Date;
  shouldUpdateEntries: boolean;
  config: any;
}

export interface ElasticSearchService {
  pushMetrics<T extends MetricItem>(
    metricItems: T[],
    shouldReplaceEntry?: boolean
  ): Promise<void>;
}
