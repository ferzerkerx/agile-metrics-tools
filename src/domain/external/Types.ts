export interface ExternalConfig {
  since: Date;
  until?: Date;
  type: string;
  srcType: string;
  inlineData: ExternalData[];
}

export class ExternalData {
  createdAt: Date;
  teamName: string;
  [type: string]: any;
}

export interface ExternalService {
  fetch(externalConfig: ExternalConfig): Promise<ExternalData[]>;
}

export interface ExternalRepository {
  csv(externalConfig: ExternalConfig): Promise<ExternalData[]>;
}
