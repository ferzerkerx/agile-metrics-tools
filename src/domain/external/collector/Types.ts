export class ExternalCollectorConfig {
  since: Date;
  until?: Date;
  type: string;
  srcType: string;
  metricType: string;
  inlineData: [];
  constructor({
    since,
    until = null,
    type,
    srcType,
    metricType,
    inlineData = null,
  }) {
    this.since = since ? new Date(since) : null;
    this.until = until ? new Date(until) : null;
    this.type = type;
    this.srcType = srcType;
    this.metricType = metricType;
    this.inlineData = inlineData;
  }
}
