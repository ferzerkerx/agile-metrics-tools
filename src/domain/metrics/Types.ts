import { ExternalCollectorConfig } from '../external/collector/Types';

export interface CollectorService {
  fetch(config: ExternalCollectorConfig): Promise<MetricItem[]>;
}

export interface MetricItem {
  id: string;
  dataType: string;
  createdAt: Date;
  teamName: string;
}
