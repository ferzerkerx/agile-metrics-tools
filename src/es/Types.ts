import { MetricItem } from '../metrics';

export interface ElasticSearchService {
  push(payload: MetricItem): Promise<any>;
}
