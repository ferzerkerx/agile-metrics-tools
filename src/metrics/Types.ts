export interface CollectorService<
  T extends CollectorConfig,
  K extends MetricItem
> {
  fetch(config: T): Promise<K[]>;
  supports(config: any): boolean;
}

export interface MetricItem {
  id: string;
  dataType: string;
  createdAt: Date;
  teamName: string;
}

export interface CollectorConfig {
  teamName: string;
}

export interface CollectorModuleFactory<
  T extends CollectorConfig,
  K extends MetricItem
> {
  collectorInstance(): CollectorService<T, K>;
  collectorConfiguration(obj: any): T;
}
