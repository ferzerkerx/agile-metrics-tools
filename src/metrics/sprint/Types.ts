export interface Statistics {
  moveBackward: number;
  moveForward: number;
  movedToDev: Date;
}
