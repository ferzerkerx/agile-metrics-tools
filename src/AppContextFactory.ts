import { AppConfig, AppContext, ModuleConfig } from './Types';
import { CollectorConfig, CollectorModuleFactory, Utils } from './metrics';
import { GithubModuleFactory } from './modules/github';
import { JiraModuleFactory } from './modules/jira';
import { JenkinsModuleFactory } from './modules/jenkins';
import { SonarqubeModuleFactory } from './modules/sonarqube';
import { ExternalModuleFactory } from './modules/external';

interface ModuleFactoryMappings {
  [type: string]: CollectorModuleFactory<any, any>;
}

export class AppContextFactory {
  static moduleFactoryMappings(): ModuleFactoryMappings {
    return {
      github: new GithubModuleFactory(),
      jira: new JiraModuleFactory(),
      jenkins: new JenkinsModuleFactory(),
      sonar: new SonarqubeModuleFactory(),
      external: new ExternalModuleFactory()
    };
  }

  static appContextForService(serviceName: string, config: string): AppContext {
    const moduleFactoryMapping: ModuleFactoryMappings = AppContextFactory.moduleFactoryMappings();
    const moduleFactory: CollectorModuleFactory<any, any> =
      moduleFactoryMapping[serviceName];

    if (!moduleFactory) {
      throw Error(`Unable to create module for service: ${serviceName}`);
    }

    const collectorService = moduleFactory.collectorInstance();

    return {
      appConfig: { indexPrefix: 'myindex', modules: [] },
      collectorsServices: [collectorService],
      collectorConfigs: [config]
    };
  }

  static async appContext(appConfig: AppConfig): Promise<AppContext> {
    const moduleFactoryMapping: ModuleFactoryMappings = AppContextFactory.moduleFactoryMappings();
    const supportedConfigs = AppContextFactory.supportedConfigs(
      appConfig,
      moduleFactoryMapping
    );

    const collectorConfigurations = await this.collectorConfigurations(
      moduleFactoryMapping,
      supportedConfigs
    );
    const collectorServices = supportedConfigs.map(moduleName => {
      return moduleFactoryMapping[moduleName.type].collectorInstance();
    });
    return {
      appConfig: appConfig,
      collectorsServices: collectorServices,
      collectorConfigs: collectorConfigurations
    };
  }

  private static supportedConfigs(
    appConfig: AppConfig,
    moduleFactories: ModuleFactoryMappings
  ): ModuleConfig[] {
    return appConfig.modules
      .map(moduleConfig => {
        if (!moduleConfig) {
          console.warn(`Unable to load ${moduleConfig}`);
          return null;
        }
        const module = moduleFactories[moduleConfig.type];
        if (!module) {
          console.warn(
            `Unable to process config for type ${moduleConfig.type}`
          );
          return null;
        }
        return moduleConfig;
      })
      .filter(moduleConfig => moduleConfig != null);
  }

  private static async collectorConfigurations(
    moduleFactories: ModuleFactoryMappings,
    supportedConfigs: ModuleConfig[]
  ): Promise<CollectorConfig[]> {
    const configContents: ModuleConfigurationDetails[] = await AppContextFactory.configurationDetails(
      supportedConfigs
    );

    const configsPerType: CollectorConfig[][] = configContents.map(
      configContent => {
        const entries = Array.isArray(configContent.entries)
          ? configContent.entries
          : [];
        return entries.map(item => {
          return moduleFactories[
            configContent.moduleConfig.type
          ].collectorConfiguration(item);
        });
      }
    );

    return Utils.flatMap(item => item, configsPerType);
  }

  private static async configurationDetails(
    supportedConfigs: ModuleConfig[]
  ): Promise<ModuleConfigurationDetails[]> {
    const configContentsPromises: Promise<
      ModuleConfigurationDetails
    >[] = supportedConfigs.map(async moduleConfig => {
      let configContents = null;
      try {
        configContents = await import(moduleConfig.configFile);
      } catch (error) {
        console.warn(`Unable to load file ${moduleConfig.configFile}`);
      }
      if (!configContents) {
        return null;
      }
      return { entries: configContents, moduleConfig: moduleConfig };
    });

    const configurationContents = await Promise.all(configContentsPromises);
    return configurationContents.filter(moduleConfig => moduleConfig != null);
  }
}

interface ModuleConfigurationDetails {
  entries: any[];
  moduleConfig: ModuleConfig;
}
