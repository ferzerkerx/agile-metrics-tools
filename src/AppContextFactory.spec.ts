import { AppContextFactory } from './AppContextFactory';
import { AppConfig, AppContext } from './Types';

function assertInvalidConfigs(appContext: AppContext) {
  expect(appContext.collectorsServices).not.toBeNull();
  expect(appContext.collectorsServices.length).toBe(1);

  expect(appContext.collectorConfigs).not.toBeNull();
  expect(appContext.collectorConfigs.length).toBe(0);
}

function givenAppConfig({ type, configFile }): AppConfig {
  return {
    indexPrefix: 'someIndexName',
    modules: [
      {
        type,
        configFile
      }
    ]
  };
}

describe('AppContextFactory', () => {
  it('should create appContext with supported module', async () => {
    const appConfig: AppConfig = {
      indexPrefix: 'someIndexName',
      modules: [
        {
          type: 'external',
          configFile: __dirname + '/test/test-external-config.json'
        }
      ]
    };
    const appContext: AppContext = await AppContextFactory.appContext(
      appConfig
    );
    expect(appContext.collectorsServices).not.toBeNull();
    expect(appContext.collectorsServices.length).toBe(1);

    expect(appContext.collectorConfigs).not.toBeNull();
    expect(appContext.collectorConfigs.length).toBe(1);
  });

  it('should create appContext without unsupported modules', async () => {
    const appConfig: AppConfig = givenAppConfig({
      type: 'some',
      configFile: '/some/test-external-config.json'
    });
    const appContext: AppContext = await AppContextFactory.appContext(
      appConfig
    );
    expect(appContext).toMatchSnapshot();
  });

  it('should create appContext without unsupported configs', async () => {
    const appConfig: AppConfig = givenAppConfig({
      type: 'external',
      configFile: __dirname + '/test/test-invalid-external-config.json'
    });
    const appContext: AppContext = await AppContextFactory.appContext(
      appConfig
    );
    assertInvalidConfigs(appContext);
  });

  it('should create appContext with not existent files', async () => {
    const appConfig = givenAppConfig({
      type: 'external',
      configFile: '/test/non-existent-file.json'
    });
    const appContext: AppContext = await AppContextFactory.appContext(
      appConfig
    );
    assertInvalidConfigs(appContext);
  });
});
