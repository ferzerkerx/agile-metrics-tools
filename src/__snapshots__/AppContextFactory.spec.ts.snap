// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppContextFactory should create appContext without unsupported modules 1`] = `
Object {
  "appConfig": Object {
    "indexPrefix": "someIndexName",
    "modules": Array [
      Object {
        "configFile": "/some/test-external-config.json",
        "type": "some",
      },
    ],
  },
  "collectorConfigs": Array [],
  "collectorsServices": Array [],
}
`;
